### PR TITLE
series failsafes

### PR DESCRIPTION
**Pull request by rwst**
second part of #793, depends on #796

### CODE REVIEW

**Comment by isuruf** [comment]
With this PR, #796 makes more sense now

**Comment by certik** [comment]
This looks good to me. +1

@rwst do you want to just merge this PR, or do you want to first merge #796?

**Comment by rwst** [comment]
@certik I would like to do it step-by-step, but don't know what your +1 refers to, the whole #798? If so, I'll just add a note as to the review here to #796.

Also a question: I tried to define the visitor in `series.h/cpp` but the compiler complained. Why must the class be declared and defined in `visitor.h`? I don't understand these template issues.

**Comment by certik** [comment]
My +1 was for the whole #798, as well ass #796 as a subset.

**Comment by certik** [comment]
As to the visitor question, it can be defined in `series.h` as far as I know. It just needs to be in the header file, due to the templates.

**Comment by isuruf** [comment]
@rwst, `visitor.h` includes almost all of the headers since it needs to know all the types which means `series.h` is also included.
If you include `series.h` in `visitor.h` imports become cyclic. `series.cpp` can include `visitor.h` without any problem.

File: symengine/visitor.h [hunk 1]
```diff
@@ -146,6 +146,65 @@ RCP<const Basic> coeff(const Basic &b, const RCP<const Basic> &x,
 
 set_basic free_symbols(const Basic &b);
 
+class NeedsSymbolicExpansionVisitor : public BaseVisitor<NeedsSymbolicExpansionVisitor, StopVisitor> {
```
Comment (isuruf) [inline symengine/visitor.h]: This can go in `series_visitor.h`

**Comment by rwst** [comment]
I moved it and also thought I would change pre to post order.

**Comment by rwst** [comment]
The last 2 commits still need review, could you please?

**Comment by isuruf** [comment]
+1.
Instead of `integer(0)`, you can use `zero` in the future.

**Comment by rwst** [comment]
Thanks.

**Pull request merged**